### PR TITLE
[AMD] Improve performance of dsv4 in high concurrency

### PR DESCRIPTION
**Pull request by kkHuang-amd**
## Motivation

DeepSeek-V4 on AMD (MI35X, HIP) under high concurrency (TP8 + DP-attention, c512) trails. Per-layer trace analysis (pure-prefill + a c512/ISL8192 mixed trace) showed the gap was **not** the heavy GEMM/MoE kernels but a few avoidable overheads in the prefill / dp-attention path:

- the attention-output (inverse) RoPE kernel was ~2x slower due to a scattered memory access pattern;
- the MoE-gather path did two redundant full-buffer (~940 MB) device-to-device copies before the gate GEMM;
- the replicated (TP1) shared expert was computed on the gathered global buffer, and only in prefill, so decode paid a ~dp_size× cost.

This PR fixes these. All are HIP/DSV4 specific; non-AMD paths are unaffected.

## Modifications

- **`deepseek_v4`**: enable, by default for DSV4 on HIP, the CK w8a8-block GEMM for the MLA projections (`set_force_ck_w8a8`) and the batched/flat RoPE kernels (`set_batched_rope`) in `__init__`. Env vars `SGLANG_FORCE_CK_W8A8` / `SGLANG_ROPE_BATCHED` still override.
- **`deepseek_v4_rope`**: new **flat-row** attention-output RoPE kernel. Instead of one program per `(token-block, head)` (tokens strided by `n_heads*head_dim` → scattered), iterate `(token, head)` pairs flattened as `row = token*n_heads + head`, with `BLOCK_ROWS` consecutive rows per program (rows only `head_dim` apart). Numerically bit-exact; replaces the per-head contig kernel.
- **`deepseek_v2` / `deepseek_v4`**: **shared-expert-local** computation (env-gated `SGLANG_DP_SHARED_EXPERT_LOCAL`, requires `SGLANG_SHARED_EXPERT_TP1`). Compute the replicated shared expert on this rank's **local** hidden (before the dp gather) and add it back to the local slice after the combine, for **both prefill and decode**. The shared expert is a per-token MLP, so this is identical to computing it on the gathered global buffer and keeping the local slice, but processes `1/dp_size` the rows — cancelling the TP1 "full-dim" decode penalty.
- **`parallel_state` / `dp_attention`**: add an optional `output=` buffer to `GroupCoordinator.all_gatherv` so NCCL gathers directly into the pre-allocated dp buffer. Removes the previous `torch.cat` + `copy_` (two full-buffer DtoD copies) in the gatherv MoE-gather path. `output=` defaults `None`, so all other callers are unchanged.

## Accuracy Tests

GSM8k 5-shot (DeepSeek-V4-Pro, TP8 + DP-attention, all opts on):

| Filter | exact_match |
|---|---|
| flexible-extract | 0.9568 ± 0.0056 |
| strict-match | 0.9575 ± 0.0056 |

(The flat-RoPE kernel is bit-exact vs the previous kernel; shared-expert-local is mathematically identical to the global computation.)

## Speed Tests and Profiling

MI35X, DeepSeek-V4-Pro, TP8 + DP-attention, ROCm, fp8 KV, single-stream, `--chunked-prefill-size 65536` (= 8192/rank), ratio 1.0.

### End-to-end c512 (this PR's net e2e gain)

Baseline = this PR's base (current `main`: already has CK GEMM + RoPE BLOCK_M + gatherv, but the old gather copies, the per-head RoPE, and no shared-expert-local), `SGLANG_DP_USE_GATHERV=1` only. After = this PR with `SGLANG_DP_USE_GATHERV=1 SGLANG_SHARED_EXPERT_TP1=1 SGLANG_DP_SHARED_EXPERT_LOCAL=1`. Total token throughput, mean of multiple runs:

| workload | before (main) | after (this PR) | gain |
|---|---:|---:|---:|
| 1k/1k c512 | 17,800 | 18,088 | **+1.6%** |
| 8k/1k c512 | 33,840 | 37,281 | **+10.2%** |

### Component level (c512 / ISL8192 / OSL32 trace, rank-0)

| region | before | after |
|---|---:|---:|
| attention-output RoPE (p50) | 168 us | **59 us** |
| MoE-gather → gate (nccl→topk) | 1011 us (358+356 DtoD + 297 GEMM) | **332 us** (gate GEMM only) |

RoPE micro-bench (real shape `[8192, 128, 64]`, cold HBM, MI35X): per-head contig ~2.2 TB/s vs **flat-row ~4.5 TB/s** (`BLOCK_ROWS=16, num_warps=1`), ~85% of the streaming ceiling.

### How to reproduce

Launch (8×MI35X, DeepSeek-V4-Pro). Feature flags for this PR are the 3 env vars below; the rest is the standard DSV4 AMD serving config:

```bash
SGLANG_USE_AITER=1 SGLANG_USE_ROCM700A=0 \
SGLANG_HACK_FLASHMLA_BACKEND=unified_kv_triton \
SGLANG_OPT_USE_FUSED_COMPRESS=true SGLANG_OPT_USE_FUSED_COMPRESS_TRITON=true \
SGLANG_OPT_USE_AITER_INDEXER=true SGLANG_FP8_PAGED_MQA_LOGITS_TORCH=1 \
SGLANG_DP_USE_GATHERV=1 SGLANG_SHARED_EXPERT_TP1=1 SGLANG_DP_SHARED_EXPERT_LOCAL=1 \
python3 -m sglang.launch_server \
  --model-path <DeepSeek-V4-Pro> --trust-remote-code \
  --tp 8 --dp 8 --enable-dp-attention --enable-prefill-delayer \
  --attention-backend dsv4 --kv-cache-dtype fp8_e4m3 \
  --page-size 256 --mem-fraction-static 0.90 --swa-full-tokens-ratio 0.15 \
  --disable-radix-cache --disable-shared-experts-fusion \
  --chunked-prefill-size 65536 --cuda-graph-max-bs 512 --max-running-requests 512 \
  --tool-call-parser deepseekv4 --reasoning-parser deepseek-v4 --port 8000
```

Accuracy (GSM8k):

```bash
lm_eval --model local-completions \
  --model_args model=<DeepSeek-V4-Pro>,base_url=http://localhost:8000/v1/completions,num_concurrent=64,max_retries=3,tokenized_requests=False \
  --tasks gsm8k --num_fewshot 5
```

Speed (c512; set `--random-input-len 1024` for 1k/1k):

```bash
python3 -m sglang.bench_serving --backend sglang-oai --base-url http://127.0.0.1:8000 \
  --model <DeepSeek-V4-Pro> --dataset-name random \
  --random-input-len 8192 --random-output-len 1024 --random-range-ratio 1.0 \
  --num-prompts 4096 --max-concurrency 512 --request-rate inf --warmup-requests 1024
```

For the **before** numbers, run the same on `main` without the three feature env vars (just `SGLANG_DP_USE_GATHERV=1`).

## Checklist

- [x] Format your code according to the Format code with pre-commit.
- [x] Provide accuracy and speed benchmark results according to Test the accuracy and Benchmark the speed.
- [x] Follow the SGLang code style guidance.

### CODE REVIEW

**Review by HaiShaw — APPROVED** [review]

**Comment by kkHuang-amd** [comment]
@amd-bot ci-status

**Review by 1am9trash — APPROVED** [review]
LGTM
